src/cdomain/value/cdomains/intDomain.ml

```diff
@@ -1351,17 +1353,85 @@ module BitfieldFunctor (Ints_t : IntOps.IntOps): SOverflow with type int_t = Int
 
   let shift_left ik a b =(top_of ik,{underflow=false; overflow=false})
 
-  let add ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
-
-  let sub ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
-
-  let mul ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
+  (*
+  add, sub and mul based on the paper 
+  "Sound, Precise, and Fast Abstract Interpretation with Tristate Numbers"
+  of Vishwanathan et al.
+  *)
 
-  let rec div ?no_ov ik x y = (top_of ik,{underflow=false; overflow=false})
+  let add ?no_ov ik (z1, o1) (z2, o2) =
+    let pv = Ints_t.logand o1 (Ints_t.lognot z1) in
+    let pm = Ints_t.logand o1 z1 in
+    let qv = Ints_t.logand o2 (Ints_t.lognot z2) in
+    let qm = Ints_t.logand o2 z2 in
+    let sv = Ints_t.add pv qv in
+    let sm = Ints_t.add pm qm in
+    let sigma = Ints_t.add sv sm in
+    let chi = Ints_t.logxor sigma sv in
+    let mu = Ints_t.logor (Ints_t.logor pm qm) chi in
+    let rv = Ints_t.logand sv (Ints_t.lognot mu) in
+    let rm = mu in 
+    let o3 = Ints_t.logor rv rm in 
+    let z3 = Ints_t.logor (Ints_t.lognot rv) rm in
+    ((z3, o3),{underflow=false; overflow=false})
+
+  let sub ?no_ov ik (z1, o1) (z2, o2) =
+    let pv = Ints_t.logand o1 (Ints_t.lognot z1) in
+    let pm = Ints_t.logand o1 z1 in
+    let qv = Ints_t.logand o2 (Ints_t.lognot z2) in
+    let qm = Ints_t.logand o2 z2 in
+    let dv = Ints_t.sub pv qv in
+    let alpha = Ints_t.add dv pm in
+    let beta = Ints_t.sub dv qm in
+    let chi = Ints_t.logxor alpha beta in
+    let mu = Ints_t.logor (Ints_t.logor pm qm) chi in
+    let rv = Ints_t.logand dv (Ints_t.lognot mu) in
+    let rm = mu in 
+    let o3 = Ints_t.logor rv rm in 
+    let z3 = Ints_t.logor (Ints_t.lognot rv) rm in
+    ((z3, o3),{underflow=false; overflow=false})
+
+  let mul ?no_ov ik (z1, o1) (z2, o2) =
+    let z1 = ref z1 in
+    let o1 = ref o1 in
+    let z2 = ref z2 in 
+    let o2 = ref o2 in
+    let z3 = ref BArith.one_mask in 
+    let o3 = ref BArith.zero_mask in 
+    for i = Size.bit ik downto 0 do 
+      if Ints_t.logand !o1 Ints_t.one == Ints_t.one then 
+        if Ints_t.logand !z1 Ints_t.one == Ints_t.one then 
+          let tmp = Ints_t.add (Ints_t.logand !z3 !o3) !o2 in 
+          z3 := Ints_t.logor !z3 tmp;
+          o3 := Ints_t.logor !o3 tmp
+        else
+          let tmp = fst (add ik (!z3, !o3) (!z2, !o2)) in 
+          z3 := fst tmp;
+          o3 := snd tmp
+      ;
```

Comment:
vll. das ";" eins rauf. 🤔

Reply:
Also das "let ... in .." Konstrukt weglassen oder was ist genau der Vorschlag? :sweat_smile: